### DISCUSSION TITLE
Can you add hibernate timer?

### DISCUSSION THREAD

**Discussion by hiensumi** [Deminder/ShutdownTimer#20] [Ideas]
Can you add hibernate timer?

**Comment by Deminder** [comment]
The option will only show up under `Shutdown > Use mode` if `CanHibernate` returns `yes`.

https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.login1.html says:

> CanPowerOff(), CanReboot(), CanHalt(), CanSuspend(), CanHibernate(), CanHybridSleep(), CanSuspendThenHibernate(), CanRebootParameter(), CanRebootToFirmwareSetup(), CanRebootToBootLoaderMenu(), and CanRebootToBootLoaderEntry() test whether the system supports the respective operation and whether the calling user is allowed to execute it. Returns one of "na", "yes", "no", and "challenge". If "na" is returned, the operation is not available because hardware, kernel, or drivers do not support it. If "yes" is returned, the operation is supported and the user may execute the operation without further authentication. If "no" is returned, the operation is available but the user is not allowed to execute the operation. If "challenge" is returned, the operation is available but only after authorization.



You can check what it returns with:
```
gdbus call --system --dest org.freedesktop.login1 --object-path /org/freedesktop/login1 --method org.freedesktop.login1.Manager.CanHibernate
```

## Idea to improve the current behavior

It could make sense to also make it easier to configure the polkit rules for this.
With a polkit rule during `Install`, maybe:
```javascript
// /etc/polkit-1/rules.d/11-dem.shutdowntimer.hibernate-{{USER}}.rules
polkit.addRule(function(action, subject) {
    if (action.id == "org.freedesktop.login1.hibernate" && subject.isInGroup("{{USER}}"))
    {
        return polkit.Result.YES;
    }
});
```
(see https://github.com/systemd/systemd/blob/8810b782a17050d7f7a870b975f09e8a690b7bea/src/login/logind-action.c#L92-L100)

But I'm not sure yet if this is really a good idea. Always allowing all possible shutdown actions seems bad.

**Reply by hiensumi** [reply]
Hibernation is supported on my machine. I have made the file like you said but nothing changed.
There are only three options (power off, suspend and restart)
![image](https://github.com/Deminder/ShutdownTimer/assets/47080203/e0a8d5f1-c524-48bb-8fd1-3df56e605c39)

**Reply by Deminder** [reply]
What does `CanHibernate()` return for you? You say hibernation is supported on your system so it should be `no` or `challenge`.

**Reply by Deminder** [reply]
Note that the hibernate option is only available since release 42, this means only for GNOME 45 and later.

**Reply by hiensumi** [reply]
![image](https://github.com/Deminder/ShutdownTimer/assets/47080203/4f595ce4-8a96-4e61-b89d-4ec0d99403f7)
I can hibernate normally in Debian 12
![image](https://github.com/Deminder/ShutdownTimer/assets/47080203/15b90b16-023f-44bd-81a1-445fe1190ec3)

**Reply by hiensumi** [reply]
I am in GNOME Shell 43.9

**Reply by Deminder** [reply]
> I am in GNOME Shell 43.9

Unfortunately, this feature is only available for GNOME 45 and later.

It would require a lot of work supporting older versions since GNOME 45 has made compatibility with older versions impossible (GNOME `≤44` ).
